dataplug-instagram/app/com/hubofallthings/dataplugInstagram/apiInterfaces/authProviders/InstagramProvider.scala

```diff
@@ -0,0 +1,145 @@
+/*
+ * Copyright (C) 2020 Dataswift Ltd - All Rights Reserved
+ * Unauthorized copying of this file, via any medium is strictly prohibited
+ * Proprietary and confidential
+ * Written by Marios Tsekis <marios.tsekis@dataswift.io>, 2 2020
+ */
+
+package com.hubofallthings.dataplugInstagram.apiInterfaces.authProviders
+
+import com.mohiva.play.silhouette.api.LoginInfo
+import com.mohiva.play.silhouette.api.util.HTTPLayer
+import com.mohiva.play.silhouette.impl.exceptions.ProfileRetrievalException
+import com.mohiva.play.silhouette.impl.providers._
+import InstagramProvider._
+import play.api.libs.json._
+
+import scala.concurrent.Future
+
+/**
+ * Base Instagram OAuth2 Provider.
+ *
+ * @see https://monzo.com/docs/#authentication
+ */
+trait BaseInstagramProvider extends OAuth2Provider {
+
+  /**
+   * The content type to parse a profile from.
+   */
+  override type Content = JsValue
+
+  /**
+   * The provider ID.
+   */
+  override val id = ID
+
+  /**
+   * Defines the URLs that are needed to retrieve the profile data.
+   */
+  override protected val urls = Map("api" -> settings.apiURL.getOrElse(API))
+
+  /**
+   * Builds the social profile.
+   *
+   * @param authInfo The auth info received from the provider.
+   * @return On success the build social profile, otherwise a failure.
+   */
+  override protected def buildProfile(authInfo: OAuth2Info): Future[Profile] = {
+    getLongLivedToken(authInfo).flatMap { refreshedTokenAuthInfo =>
+      val url = httpLayer.url(urls("api").format(refreshedTokenAuthInfo.accessToken)).addQueryStringParameters(("access_token", refreshedTokenAuthInfo.accessToken))
+      url.get().flatMap { response =>
+        val json = response.json
+        (json \ "code").asOpt[Int] match {
+          case Some(code) if code != 200 =>
+            val errorType = (json \ "error_type").asOpt[String]
+            val errorMsg = (json \ "error_message").asOpt[String]
+
+            throw new ProfileRetrievalException(SpecifiedProfileError.format(id, code, errorType, errorMsg))
+          case _ => profileParser.parse(json, refreshedTokenAuthInfo)
+        }
+      }
+    }
+  }
+
+  private def getLongLivedToken(authInfo: OAuth2Info): Future[OAuth2Info] = {
+    val clientSecret = sys.env("API_CLIENT_SECRET")
```

Comment:
Direct access to the environment variables is discouraged. Use configuration file to pass environment variables.